tests/Console/E2ETest.php

```diff
@@ -0,0 +1,243 @@
+use Infection\Finder\ComposerExecutableFinder;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Console\Input\ArgvInput;
+use Symfony\Component\Console\Output\BufferedOutput;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Process\Process;
+
+class E2ETest extends TestCase
+{
+    const EXPECT_ERROR = 1;
+    const EXPECT_SUCCESS = 0;
+
+    private $cwd;
+
+    /**
+     * @var \Composer\Autoload\ClassLoader|null
+     */
+    private $previousLoader;
+
+    protected function setUp()
+    {
+        // Without overcommit this test fails with `proc_open(): fork failed - Cannot allocate memory`
+        if (strpos(PHP_OS, 'Linux') === 0 &&
+            is_readable('/proc/sys/vm/overcommit_memory') &&
+            file_get_contents('/proc/sys/vm/overcommit_memory') == 2) {
+            $this->markTestSkipped('This test needs copious amounts of virtual memory. It will fail unless it is allowed to overcommit memory.');
+        }
+
+        // E2E tests usually require to chdir to their location
+        // Hence we would need to go back to this dir
+        $this->cwd = getcwd();
+    }
+
+    /**
+     * Longest test: runs under about 160-200 sec
+     *
+     * To be run with:
+     *
+     * php -dmemory_limit=128M vendor/bin/phpunit --group=large
+     *
+     * @group e2e
+     * @large
+     */
+    public function test_it_runs_on_itself()
+    {
+        if (ini_get('memory_limit') === '-1') {
+            $this->markTestSkipped(implode("\n", [
+                'Refusing to run Infection on itself with no memory limit set: it is dangerous.',
+                'To run this test with a memory limit set please use:',
+                'php -dmemory_limit=128M vendor/bin/phpunit --group=large',
+            ]));
+        }
+
+        $output = $this->runInfection(self::EXPECT_SUCCESS, [
+            '--test-framework-options="--exclude-group=e2e"',
+        ]);
+
+        $this->assertRegExp('/\d+ mutations were generated/', $output);
+        $this->assertRegExp('/\d{2,} mutants were killed/', $output);
+    }
+
+    /**
+     * @group e2e
+     */
+    public function test_it_runs_configure_command_if_no_configuration()
+    {
+        chdir('tests/Fixtures/e2e/Unconfigured/');
+
+        $output = $this->runInfection(self::EXPECT_ERROR);
+
+        $this->assertContains(ConfigureCommand::NONINTERACTIVE_MODE_ERROR, $output);
+    }
+
+    /**
+     * @dataProvider e2eTestSuiteDataProvider
+     * @group e2e
+     */
+    public function test_it_runs_an_e2e_test_with_success(string $e2eTestName, string $fullPath)
+    {
+        $this->runOnE2EFixture($fullPath);
+    }
+
+    public function e2eTestSuiteDataProvider(): \Generator
+    {
+        $directories = Finder::create()
+            ->depth('== 0')
+            ->in(__DIR__ . '/../Fixtures/e2e/')
+            ->directories();
+
+        foreach ($directories as $dirName) {
+            if (file_exists($dirName . '/run_tests.bash')) {
```

Comment:
Probably, it would be better to create an array of the tests we want to skip her, because just checking custom `run_tests.bash` file existence is too dirty check. It skips too many cases, even if the test is not "very slow to fail".

what do you think?

Reply:
We assume there's always an `expected-output.txt` present in the test directory. When a non-standard `run_tests.bash` is used we cannot be sure there's that file. I think it is better to follow a convention than add another knob to configure a test suite. One more thing to remember, and one more thing to overlook. If we can do without it, why not?

Reply:
We can do something like 

```php
if (file_exists('expected-output.txt')) {
    $this->assertFileEquals('expected-output.txt', 'infection-log.txt', ...);
}
```

then.

What I want to say is when we skip so many e2e tests (8 of 15 atm), we lose code coverage.

Reply:
If we're not asserting anything, like against known `expected-output.txt`, for what kind of coverage will be these tests for then? We already had to lower MSI requirement from 86% to 72%.

OK, we'll be testing against exit code, but these tests are testing against it too. Not sure if that's going to give us any coverage bump, this looks very much redundant.

Reply:
> If we're not asserting anything, like against known expected-output.txt, for what kind of coverage will be these tests for then?

Let's take `Ignore_MSI_Zero_Mutations` and `Test_Frameworks`: both folders has custom `run_tests.bash`, but they also have `expected-output.txt`. Why do we skip them? Just because some other test folder with custom `run_tests.bash` *probably* does not contain `expected-output.txt`? That's strange as for me.

In this case it's better to replace `if (file_exists($dirName . '/run_tests.bash')) { continue }` with `if (! file_exists($dirName . '/expected-output.txt')) { continue }`

Reply:
That'd be a bummer if we ever happen to add a *negative* test which would also have `expected-output.txt`. You can't say from `expected-output.txt` alone if you're looking at positive or negative test. Yet from missing run_tests.bash you can tell for certain - this a positive test, that should not err.

Reply:
If we'd like you say, a present `expected-output.txt` would mean that this *must* be a positive test. And if you would ever need to design a negative test, you would be forced to use a different name, like `expected-output2.txt`. That's silly, don't you think?

Reply:
Yes, that's silly. But it is not very good to skip good e2e tests just because they have custom `run_tests.bash` while they are positive and have `expected-output.txt`. 

Any new positive test with `expected-output.txt` but with custom `run_tests.bash` will be never run by PHPUnit - this is the biggest concern I have.

Reply:
I'm not as concerned here because a *custom* test probably will require a *proper customized* procedure to run, which is [a piece of cake within the current architecture](https://github.com/infection/infection/pull/285/files#diff-c50be878dbd21bc70fc3c4da7deedccfR80). It is the right thing to do to a add a specialized testing procedure for a specialized non-standard test, don't you think?

Reply:
So my point is: 

- If you're adding a standard test, no worries everything is taken care for you.
- If you're adding a custom test, please be a good fellow and code a custom procedure for it.